### DISCUSSION TITLE
umijs 内置插件有没有可能支持 oidc 登录

### DISCUSSION THREAD

**Discussion by AlbertXiaoPeng** [umijs/umi#11063] [Ideas]
目前大多数登录为两种：
1. 普通账号密码换取token
2. oidc，或者也可以说是oauth2.0 单点登录这种

普通登录直接正常api请求就可以了，有没有可能后续增加oidc插件在内置中

毕竟项目比较适合中台管理或者后台管理，一般公司多个站点不可能每个站点登录一次，单点登录比较好。
但是自己写的在一些升级或者其他时间会出现一些问题。
比较建议能将oidc集成到umijs的插件中，可以按需取用么

**Comment by xiaohuoni** [comment]
有个小问题，oidc 登录有通用性嘛？

**Reply by AlbertXiaoPeng** [reply]
有啊，OAuth2.0是有标准的，所以完全可以只支持标准版本的就行
我们当初使用vite开发时用的是 `redux-oidc`和`oidc-client`包来实现的标准版本的登录
可以参考以下：
- http://www.rfcreader.com/#rfc6749
- https://oauth.net/2/
- [阮一峰-理解OAuth 2.0](https://www.ruanyifeng.com/blog/2014/05/oauth_2_0.html)

**Reply by xiaohuoni** [reply]
我不是这个意思，我是说落地到实际项目的时候，比如你上面提到的，你用的是  redux-oidc和oidc-client 其他人用的可能就不是了，包括我们内部我就对接过“很奇怪”的一些

**Reply by xiaohuoni** [reply]
坐等yingci老师回复

**Comment by fz6m** [comment]
我理解 OAuth 2 有两层。

第一层是规范和流程层。包括你们的 token 约定，是否遵循 `Bearer ${token}` 的请求 header ，如何存储（比如在 local storage 或 indexdb 、cookie 等）；在登录页面上，是 sso 单点登录，还是单系统登录，公司有没有 cookie 权限历史债务，怎么区分角色和权限点的，怎么一次登录完成多域名加载权限的，等等，这些都是你们根据公司自己的基础建设、历史、现状来约定的。在框架层面上没法强制约定你必须遵循什么规范和流程。

第二层是产品层。在这一层，我们通俗来讲就是在你的产品上可以支持其他公共账号的授权 OAuth 登录，比如用 google / ms 账号登录你们的系统，这种情况一般是需要你们内部做授权关联的设计和流程打通，或者选择一些 sdk ，比如 @react-oauth/google 、jaredhanson/passport ，或是一些收费聚合的第三方登录平台 SDK ；同样这也是开发者的自由，在框架层我们没法做强约定。

框架这个层面应该只做最通用的东西，具体细化到实际场景，要和公司业务挂钩，这个时候你可以做自己的 umi 插件，生成符合你们内部流程的临时文件和导出对应的方法完成需求。
